rp2040-hal/src/pwm.rs

```diff
@@ -1,3 +1,312 @@
 //! Pulse Width Modulation (PWM)
-// See [Chapter 4 Section 5](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
-// TODO
+//!
+//! To access the PWM pins you must call the 'split' method on the PWM. This will return a
+//! `_____` struct with access to each PWM pin:
+//!
+//! ```rust
+//! use rp2040_hal::prelude::*;
+//! let mut pac = rp2040_pac::Peripherals::take().unwrap();
+//! let pins = pac.PWM.split(pac.PADS_BANK0, pac.IO_BANK0, &mut pac.RESETS);
+//! ```
+//!
+//! Once you have the PWM pins struct, you can take individual pins and configure them:
+//!
+//! ```rust
+//! let mut pwm_pin = pin.pwm0.default_config();
+//! pwm_pin.set_duty(32767);
+//! pwm_pin.enable();
+//! ```
+//!
+//! The following configuration options are also available:
+//!
+//! ```rust
+//! pin.pwm0.min_config(); // Doesn't change config settings that could be shared between pins
+//!
+//! pwm_pin.get_duty();
+//! pwm_pin.get_max_duty();
+//!
+//! pwm_pin.set_ph_correct().unwrap(); // Run in phase correct mode
+//! pwm_pin.clr_ph_correct().unwrap(); // Don't run in phase correct mode
+//!
+//! pwm_pin.set_div_int(value: u8).unwrap(); // To set integer part of clock divider
+//! pwm_pin.set_div_frac(value: u8).unwrap(); // To set fractional part of clock divider
+//!
+//! pwm_pin.set_inv().unwrap(); // Invert the output
+//! pwm_pin.clr_inv().unwrap(); // Don't invert the output
+//!
+//! pwm_pin.set_top(value: u16).unwrap(); // To set the TOP register 
+//!
+//! pwm_pin.divmode_div().unwrap(); // Default divmode. Counts up at a rate dictated by div.
+//! pwm_pin.divmode_level().unwrap(); // These 3 divmodes can be used with a PWM B pin to read PWM inputs.
+//! pwm_pin.divmode_rise().unwrap();
+//! pwm_pin.divmode_fall().unwrap();
+//! ```
+//!
+//! default_config() sets ph_correct to false, the clock divider to 1, does not invert the output, sets top to 65535, and resets the counter.
+//! min_config() leaves those registers in the state they were before it was called (Careful, this can lead to unexpected behavior)
+//! It's recommended to only call min_config() after calling default_config() on a pin that shares a PWM block.
+
+#[doc = "Trait to be implemented on the HAL, containing methods to extract PWM pins"]
+pub trait PwmExt<PADS, IO> {
+    #[doc = "The type that PWM pins are stored in"]
+    type Parts;
+
+    #[doc = "Method that extracts PWM pins into HAL objects"]
+    fn split(self, pads: PADS, io: IO, reset: &mut rp2040_pac::RESETS) -> Self::Parts;
```

Comment:
Following the [C-CTOR recommendation](https://docs.rust-embedded.org/book/design-patterns/hal/predictability.html) in the Rust Embedded book, we shouldn't have a `split` method, and instead, define a type we can call `new` on.